src/main/java/com/epam/aidial/core/limiter/RateLimiter.java

```diff
@@ -129,11 +147,72 @@ public String toString() {
         }
     }
 
-    private record Entity(String id, List<String> roles, String name, boolean user) {
+    @Data
+    private static class Entity {
+
+        private final String id;
+        private final List<String> roles;
+        private final String name;
+        private final boolean user;
+        private final Map<String, CallInfo> spanIdToCallInfo = new HashMap<>();
```

Comment:
Please move the logic related to token tracking to a separate class. e.g. TokenTracker